synapse/handlers/sync.py

```diff
@@ -183,6 +183,20 @@ def __bool__(self) -> bool:
         return bool(self.join or self.invite or self.leave)
 
 
+@attr.s(slots=True, frozen=True, auto_attribs=True)
+class MembershipChangesSummary:
```

Comment:
I'm not sure if using a class here instead of the 4-tuple is much clearer.

Reply:
I'm glad you agree---I wasn't convinced either. I'll revert https://github.com/matrix-org/synapse/commit/0e2c6f6fcd13237514896d64af8b039e2cf4d9f4.

Reply:
... reverted in c38478998.